ktor-utils/common/src/io/ktor/util/Text.kt

```diff
@@ -39,6 +39,66 @@ inline fun String.chomp(separator: String, onMissingDelimiter: () -> Pair<String
     }
 }
 
+/**
+ * Does the same as the regular [toLowerCase] except that locale-specific rules are not applied to ASCII characters
+ * so latin characters are converted by the original english rules.
+ */
+@InternalAPI
+fun String.toLowerCasePreservingASCIIRules(): String {
+    val firstIndex = indexOfFirst {
+        toLowerCasePreservingASCII(it) != it
+    }
+
+    if (firstIndex == -1) {
+        return this
+    }
+
+    val original = this
+    return buildString(length) {
+        append(original, 0, firstIndex)
+
+        for (index in firstIndex .. original.lastIndex) {
+            append(toLowerCasePreservingASCII(original[index]))
+        }
+    }
+}
+
+/**
+ * Does the same as the regular [toUpperCase] except that locale-specific rules are not applied to ASCII characters
+ * so latin characters are converted by the original english rules.
+ */
+@InternalAPI
+fun String.toUpperCasePreservingASCIIRules(): String {
+    val firstIndex = indexOfFirst {
+        toUpperCasePreservingASCII(it) != it
+    }
+
+    if (firstIndex == -1) {
+        return this
+    }
+
+    val original = this
+    return buildString(length) {
+        append(original, 0, firstIndex)
+
+        for (index in firstIndex .. original.lastIndex) {
+            append(toUpperCasePreservingASCII(original[index]))
+        }
+    }
+}
+
+private fun toLowerCasePreservingASCII(ch: Char): Char = when (ch) {
+    in 'A' .. 'Z' -> ch + 32
```

Comment:
Do we actually need here two different branches for `A` .. `Z` and else?

Reply:
Yes, this is the point: we always do lowercase manually for A-Z. The else block does `toLowerCase` that for letters in range A-Z may produce characters other than a-z.

Reply:
This is not obvious but it may happen due to locale-specific rules